feat: prove an extensionality lemma for bitvectors

alexkeizer opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

alexkeizer commented on Mathlib/Data/BitVec/Lemmas.lean:
```diff
@@ -105,4 +105,73 @@ theorem ofFin_toFin {n} (v : BitVec n) : ofFin (toFin v) = v := by
+@[simp] lemma getLsb'_xor : (x ^^^ y).getLsb' i = (xor (x.getLsb' i) (y.getLsb' i)) := by
+  simp only [getLsb', getLsb, toNat_xor, testBit_xor]
+
+@[simp] lemma getLsb'_zero : getLsb' 0#w i = false := by
```
Do you have a strong preference for `0` as the simp normal form? The relevant simp-lemma has a note "TODO: This needs more usage data to assess which direction the simp should go.", so I assume std would be receptive to a PR which reverses it.

eric-wieser commented on Mathlib/Data/BitVec/Lemmas.lean:
```diff
@@ -218,4 +231,58 @@ instance : CommSemiring (BitVec w) :=
+    ∀ (i : ℕ), x.getLsb i = y.getLsb i := by
+  simp only [getLsb, testBit]
+  intro i
+  by_cases hi : i < w
```
I think `cases lt_or_le i w` gives a nicer proof.

alexkeizer replied:
I changed the proof to use `lt_or_le`

eric-wieser commented on Mathlib/Data/BitVec/Lemmas.lean:
```diff
@@ -218,4 +231,58 @@ instance : CommSemiring (BitVec w) :=
+section
+variable (x y : BitVec w) (i : Fin w)
+
+@[simp] lemma getLsb'_and : (x &&& y).getLsb' i = (x.getLsb' i && y.getLsb' i) := by
```
Can you add `not` or whatever `~~~` is called too?

alexkeizer replied:
done

tobiasgrosser commented:
I paired with @bollu, and we introduced `pred_bit0` to restating `pred_bit_false` with lhs in simplified form. This fixes a CI error.

alexkeizer commented:
After having another look, the "Distributivity of `Std.BitVec.toNat`" section seemed incomplete, so I've added relevant lemmas about multiplication, subtraction, negation and `Nat.cast` as well, and added a note that `toNat_add` and `toNat_zero` are already in Std

tobiasgrosser commented:
@eric-wieser, this one could also benefit from a review.

eric-wieser commented on Mathlib/Data/BitVec/Lemmas.lean:
```diff
@@ -218,4 +244,62 @@ instance : CommSemiring (BitVec w) :=
+  simp only [testBit_toNat]
+  exact getLsb_eq_of_getLsb' h
+
+theorem getMsb'_eq_getLsb' (x : BitVec w) (i : Fin w) :
+    x.getMsb' i = x.getLsb' i.rev := by
```
```suggestion
theorem getLsb'_rev (x : BitVec w) (i : Fin w) :
    x.getLsb' i.rev = x.getMsb' i := by
```
and then probably we should have a `getMsb'_rev` lemma too.

tobiasgrosser replied:
A added a placeholder for getMsb'_rev and marked this as waiting-for-author.

tobiasgrosser commented:
I updated this PR to master. It merges cleanly after updating the use of a deprecated lemma and had all review comments addressed. Is this ready to go?

ChrisHughes24 commented on Mathlib/Data/BitVec/Lemmas.lean:
```diff
@@ -218,4 +244,66 @@ instance : CommSemiring (BitVec w) :=
+    rw [Nat.shiftRight_eq_zero_of_lt (this x), Nat.shiftRight_eq_zero_of_lt (this y)]
+
+/-- If two bitvectors agree on all bits, then they are equal. See also `Std.BitVec.ext_msb` -/
+@[ext]
+theorem ext_lsb {x y : BitVec w} (h : ∀ i, x.getLsb' i = y.getLsb' i) : x = y := by
```
I wouldn't put the extensionality attribute on this lemma. I don't think this is the first thing people expect when they think of an extensionality for bitvectors, and the `ext` tactic should usually use the first thing people expect. I don't think anything is the first thing people expect for bitvectors to be honest.

eric-wieser replied:
This is the first thing I'd expect; "two bitvectors are equal if they have the same bits; but maybe this is like `Complex.ext`, which is the "obvious" ext lemma, but often you don't actually want it.

ChrisHughes24 replied:
I don't feel that strongly about this comment, I'll happily leave the `ext` attribute there.

tobiasgrosser replied:
I would prefer to keep it as `ext`. It seems quite natural for me. I resolve this for now as @ChrisHughes24 does not seem to feel strongly here.

ChrisHughes24 commented on Mathlib/Data/Nat/Bitwise.lean:
```diff
@@ -296,6 +296,30 @@ theorem testBit_two_pow (n m : ℕ) : testBit (2 ^ n) m = (n = m) := by
 #align nat.test_bit_two_pow Nat.testBit_two_pow
 
+@[simp]
+lemma pred_bit0 {x : ℕ} (h : x > 0) : pred (bit0 x) = bit true (pred x) := by
```
Is `bit true` equal to `bit1`? If so I think that might be the most sensible statement.

alexkeizer replied:
They are equal, yes. I tried to avoid `bit1` sering as it is deprecated. Then again, it is still the simp normal form (which is what forced me to use `bit0`), so it makes sense to change.

alexkeizer replied:
We have the applied version as [bit_false](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Nat/Bitwise.html#Nat.bit_false) and bit_true, would those suffice?

bollu replied:
That works better, I have pushed to use these lemmas.

alexkeizer replied:
Perfect, then I'm marking this as resolved, since we made the suggested change

tobiasgrosser commented:
I updated this to master again with all comments addressed. This is now ready for review again.

linesthatinterlace commented:
I've just reviewed this myself and it's looking pretty good. Unless there's something obvious I've missed I would say that a maintainer could feel confident in this patch.

alexkeizer commented on Mathlib/Data/BitVec/Lemmas.lean:
```diff
@@ -217,4 +243,66 @@ instance : CommSemiring (BitVec w) :=
+## Extensionality
+-/
+
+/-- If two bitvectors agree on all in-bound bits, then they agree on all bits -/
+theorem getLsb_eq_of_getLsb' {x y : BitVec w} (h : ∀ (i : Fin w), x.getLsb' i = y.getLsb' i) :
+    ∀ (i : ℕ), x.getLsb i = y.getLsb i := by
+  simp only [getLsb, testBit]
+  intro i
+  cases' lt_or_le i w with hi hi
+  · exact h ⟨i, hi⟩
+  · have (z : BitVec w) : z.toNat < 2 ^ i :=
+      Nat.lt_of_lt_of_le z.toNat_lt (pow_le_pow_right (le_succ 1) hi)
+    rw [Nat.shiftRight_eq_zero_of_lt (this x), Nat.shiftRight_eq_zero_of_lt (this y)]
+
+/-- If two bitvectors agree on all bits, then they are equal. See also `Std.BitVec.ext_msb` -/
+@[ext]
+theorem ext_lsb {x y : BitVec w} (h : ∀ i, x.getLsb' i = y.getLsb' i) : x = y := by
+  apply toNat_inj.mp
+  apply Nat.eq_of_testBit_eq
+  simp only [testBit_toNat]
+  exact getLsb_eq_of_getLsb' h
```
The first is used as part of the proof of the second. Of course, the second lemma supersedes the first, so we could just inline it, but this seemed a nice way to break up the proof. If you're concerned about API surface, we could make first lemma `private`?

alexkeizer commented on Mathlib/Data/Nat/Bitwise.lean:
```diff
@@ -492,4 +516,9 @@ lemma append_lt {x y n m} (hx : x < 2 ^ n) (hy : y < 2 ^ m) : y <<< n ||| x < 2
   · apply lt_of_lt_of_le hx <| pow_le_pow_right (le_succ _) (le_add_right _ _)
 
+theorem shiftRight_eq_zero_of_lt {x y} (h : x < 2^y) :
+    x >>> y = 0 := by
```
I've changed it to be an iff

alexkeizer commented on Mathlib/Data/BitVec/Lemmas.lean:
```diff
@@ -197,6 +197,32 @@ lemma toFin_ofNat (n : ℕ) :
+  simp only [Neg.neg, BitVec.neg, BitVec.sub_eq, toNat_sub, ofNat_eq_ofNat, toNat_zero, zero_add]
+
+@[simp] lemma toNat_natCast (n : ℕ) : toNat (n : BitVec w) = n % 2 ^ w := by
+  rw [toNat, toFin_natCast, Fin.coe_ofNat_eq_mod]
```
The general pattern is to use `toNat` to transform a goal/statement about bitvectors into a statement about natural numbers instead. It seems useful to have this be done with just `simp`, and there's no way to do that without introducing the modulus.
However, in the interest of getting this merged quicker, I have removed the simp attribute from these lemmas

alexkeizer commented:
The latest review comments have been addressed, @YaelDillies could you have a look to see if you're happy with our response?

ericrbg commented on Mathlib/Data/BitVec/Lemmas.lean:
```diff
@@ -197,6 +197,32 @@ lemma toFin_ofNat (n : ℕ) :
+@[simp] lemma toNat_or  : (x ||| y).toNat = x.toNat ||| y.toNat := rfl
+@[simp] lemma toNat_xor : (x ^^^ y).toNat = x.toNat ^^^ y.toNat := rfl
+
+/- `Std.BitVec.toNat_add` and `Std.BitVec.toNat_zero` already exists in Std -/
```
`toNat_zero` is deprecated - I wouldn't mention it

ericrbg commented on Mathlib/Data/BitVec/Lemmas.lean:
```diff
@@ -197,6 +197,32 @@ lemma toFin_ofNat (n : ℕ) :
+lemma toNat_neg : (-x).toNat = (2 ^ w - x.toNat) % 2 ^ w := by
+  simp only [Neg.neg, BitVec.neg, BitVec.sub_eq, toNat_sub, ofNat_eq_ofNat, toNat_zero, zero_add]
+
+lemma toNat_natCast (n : ℕ) : toNat (n : BitVec w) = n % 2 ^ w := by
```
Isn't this https://leanprover-community.github.io/mathlib4_docs/Std/Data/BitVec/Lemmas.html#Std.BitVec.toNat_ofNat

alexkeizer replied:
Not quite, the linked lemmas talks about `BitVec.ofNat` on the lhs, whereas this lemmas talks about `Nat.cast`. These two expressions are def-eq, though, so they are definitely related.

I don't think we immediately need this result for the main contribution of this PR, so I'd be happy to remove it if the phrasing of the lemma is controversial.

ericrbg replied:
If they're defeq, I don't see why we should have both - `Nat.cast` should be the only spelling, imo

ericrbg commented on Mathlib/Data/BitVec/Lemmas.lean:
```diff
@@ -217,4 +243,92 @@ instance : CommSemiring (BitVec w) :=
+@[simp] lemma getMsb'_ofNat_zero : getMsb' 0#w i = false := by
+  simp only [← getLsb'_rev, getLsb'_ofNat_zero]
+
+proof_wanted getMsb'_negOne : getMsb' (-1) i = true
```
Maybe leave a proof outline based on `getLsb'_negOne` commented

alexkeizer replied:
The proof is just `simp only [← getLsb'_rev, getLsb'_negOne]`. I've added a comment with this!

alexkeizer commented:
I've merged master, so that the [recent STD pr](https://github.com/leanprover/std4/pull/558) that deprecated `toNat_zero` is also part of this branch, and addressed all comments. @ericrbg, please check that you're happy with our response!

ericrbg commented:
I think I have to say it on its own line

maintainer merge

semorrison commented:
@alexkeizer, @tobiasgrosser, apologies for the delay on this. We've realised we need to make BitVec an urgent priority, and are planning on working on this in Std starting right now. :-) If it's okay, we will adapt this material to Std (in fact some of it has already independently arrived in the meantime). We would really like to keep the BitVec material coherent (ideally no one will feel any need to have anything about `BitVec` in Mathlib)! Don't hesitate to contact me and @joehendrix about BitVec related things!

eric-wieser commented:
> ideally no one will feel any need to have anything about `BitVec` in Mathli

I don't see the ring instance moving to Std any time soon, or even ever.

linesthatinterlace commented:
I've been working on a bunch of stuff previously which would ideally use BitVec (I haven't because it was previously unstable) but which would have to fit into MathLib if it fitted in anywhere because of other things it uses and proves (it's about permutations on bitvectors). I am a bit nervous about the idea of these things all moving into Std, where they become a lot more irrevocable and hard to change.

eric-wieser commented:
Given this has already occupied a bunch of review time and been approved, I would be inclined to merge it; we can always remove the bits that end up upstreamed to Std later, but this way we don't lose the bits that aren't upstreamed.

linesthatinterlace commented:
I certainly think it would be good to merge it.

semorrison commented:
As I understand, material equivalent to this in now in Std's review queue, and further BitVec development will happen there.

linesthatinterlace commented:
That still isn't an argument against merging it into this project now.

tobiasgrosser commented:
I tend to agree with @semorrison. Let's merge as much BitVector work as possible into std4 and keep anything that strictly requires mathlib in mathlib. @alexkeizer is there anything in this PR that is not covered by @semorrison's std4 PR?

tobiasgrosser commented:
Fortunately, a lot of the review work that went into this PR will live on in std4.

linesthatinterlace commented:
I think my concern is that from my point of view as a mathlib contributer, going into standard is like being carved into stone, and we already have issues where std makes choices that affect things downstream that we can't fix. This being said I think you probably have the design right.

tobiasgrosser commented:
I guess that is desired as many people are starting to build on BitVec, so agreeing on a standard might outweigh the benefit of not agreeing whichever the choice is. That having said, std itself is still evolving and definitions for BitVec keep improving. I am sure the std maintainers would appreciate comments, experience reports, ...

Also, the PR for std is currently up for review. Feel invited to contribute there:
https://github.com/leanprover/std4/pull/596

alexkeizer commented:
I'm closing this as it has now been superseded by https://github.com/leanprover/std4/pull/596
